### PR TITLE
Improve playground requests and state management

### LINKED ISSUE #49: Refactor requests
# Motivation

Right now we have some code duplication with the requests and they can be extracted to some different location. We are using `useSWR` at the moment which is more focused on GET requests. We might want to look into `react-query` to handle both `POST` and `GET` requests in a nice way.

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
### Motivation and Context

The requests and the state management kind of got hard to manage. I'm refactoring the Playground in order to make it more maintainable and the requests reusable.

### Description

- Use `react-query` with hooks instead of `swr` and create hooks for all endpoints
- Move the base info into a globally managed state so we don't have to pass it through all the children
- Some minor changes I found along the way

### Steps for Testing

Test if the playground works as usual. I basically refactored all requests and state management so everything needs testing.

### CODE REVIEW

**Comment by FelixTJDietrich** [comment]
~~Converting to draft to fix some errors, first. Will make it ready for review soon~~

Ready for review again, removed some unnecessarily thrown errors

**Comment by pal03377** [comment]
Note: I've added a small change to fix the CI build

File: playground/src/components/details/exercise_detail/text.tsx [hunk 1]
```diff
@@ -7,9 +7,7 @@ type TextExerciseDetailProps = {
   exercise: TextExercise;
 };
 
-export default function TextExerciseDetail({
-  exercise,
-}: TextExerciseDetailProps) {
+export default function TextExerciseDetail({ exercise }: TextExerciseDetailProps) {
```
Comment (pal03377) [inline playground/src/components/details/exercise_detail/text.tsx]: I think this is clearer:
```suggestion
export default function TextExerciseDetail({ exercise }: { exercise: TextExercise }) {
```
(and removing the `TextExerciseDetailProps` above)

File: playground/src/components/module_requests/get_config_schema.tsx [hunk 2]
```diff
@@ -72,18 +32,16 @@ export default function GetConfigSchema({
         signature of all other decorators to provide the configuration to the
         function.
       </p>
-      <ModuleResponseView response={response} />
+      <ModuleResponseView response={data || (error && error.asModuleResponse ? error.asModuleResponse() : undefined)} />
```
Comment (pal03377) [inline playground/src/components/module_requests/get_config_schema.tsx]:
```suggestion
      <ModuleResponseView response={data ?? (error?.asModuleResponse ? error.asModuleResponse() : undefined)} />
```

File: playground/src/components/module_requests/index.tsx [hunk 3]
```diff
@@ -8,22 +8,18 @@ import RequestFeedbackSuggestions from "@/components/module_requests/request_fee
...
 export type ModuleRequestProps = {
-  mode: Mode;
-  athenaUrl: string;
-  athenaSecret: string;
   module: ModuleMeta;
-  moduleConfig: any;
 };
 
-export default function ModuleRequests(props: ModuleRequestProps) {
+export default function ModuleRequests({ module }: ModuleRequestProps) {
```
Comment (pal03377) [inline playground/src/components/module_requests/index.tsx]:
```suggestion
export default function ModuleRequests({ module }: { module: ModuleMeta }) {
```

File: playground/src/components/module_requests/request_feedback_suggestions.tsx [hunk 4]
```diff
@@ -150,27 +105,34 @@ export default function RequestFeedbackSuggestions({
           </div>
         </>
       )}
-      <ModuleResponseView response={response}>
+      <ModuleResponseView response={response || (error && error.asModuleResponse ? error.asModuleResponse() : undefined)}>
```
Comment (pal03377) [inline playground/src/components/module_requests/request_feedback_suggestions.tsx]:
```suggestion
      <ModuleResponseView response={response ?? (error?.asModuleResponse ? error.asModuleResponse() : undefined)}>
```

File: playground/src/components/module_requests/request_submission_selection.tsx [hunk 5]
```diff
@@ -143,38 +74,43 @@ export default function SelectSubmission({
...
       {exercise && (
         <div className="space-y-1 mt-2">
-          <ExerciseDetail exercise={exercise} mode={mode} />
-          <SubmissionList exercise={exercise} mode={mode} />
+          <ExerciseDetail exercise={exercise} />
+          <SubmissionList exercise={exercise} />
         </div>
       )}
-      <ModuleResponseView response={response}>
+      <ModuleResponseView response={response || (error && error.asModuleResponse ? error.asModuleResponse() : undefined)}>
```
Comment (pal03377) [inline playground/src/components/module_requests/request_submission_selection.tsx]:
```suggestion
      <ModuleResponseView response={response ?? (error?.asModuleResponse ? error.asModuleResponse() : undefined)}>
```

File: playground/src/components/module_requests/send_feedback.tsx [hunk 6]
```diff
@@ -268,53 +131,58 @@ export default function SendFeedback({
           )}
         </>
       )}
-      {responses?.map((response, i) => (
-        <ModuleResponseView key={i} response={response} />
-      ))}
+      {responses?.map((response, i) => (<ModuleResponseView key={i} response={response} />))}
+      {error && error.asModuleResponse && (<ModuleResponseView response={error.asModuleResponse()} />)}
```
Comment (pal03377) [inline playground/src/components/module_requests/send_feedback.tsx]:
```suggestion
      {error?.asModuleResponse && (<ModuleResponseView response={error.asModuleResponse()} />)}
```

File: playground/src/components/module_requests/send_submissions.tsx [hunk 7]
```diff
@@ -106,36 +39,41 @@ export default function SendSubmissions({
...
       {exercise && (
         <div className="space-y-1 mt-2">
-          <ExerciseDetail exercise={exercise} mode={mode} />
-          <SubmissionList exercise={exercise} mode={mode} />
+          <ExerciseDetail exercise={exercise} />
+          <SubmissionList exercise={exercise} />
         </div>
       )}
-      <ModuleResponseView response={response} />
+      <ModuleResponseView response={response || (error && error.asModuleResponse ? error.asModuleResponse() : undefined)} />
```
Comment (pal03377) [inline playground/src/components/module_requests/send_submissions.tsx]:
```suggestion
      <ModuleResponseView response={response ?? (error?.asModuleResponse ? error.asModuleResponse() : undefined)} />
```

File: playground/src/components/selectors/submission_select.tsx [hunk 8]
```diff
@@ -64,7 +55,7 @@ export default function SubmissionSelect({
               ✨ All submissions
             </option>
           )}
-        {data.map((sub: Submission) => {
+        {data && data.map((sub: Submission) => {
```
Comment (pal03377) [inline playground/src/components/selectors/submission_select.tsx]:
```suggestion
        {data?.map((sub: Submission) => {
```

File: playground/src/hooks/athena/send_feedback.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ * mutate({ exercise, submission, feedback });
+ *
+ * @param options The react-query options.
+ */
+export function useSendFeedback(
+  options: Omit<
+    UseMutationOptions<ModuleResponse | undefined, AthenaError, { exercise: Exercise; submission: Submission; feedback: Feedback; }>,
+    "mutationFn"
+  > = {}
+) {
```
Comment (pal03377) [inline playground/src/hooks/athena/send_feedback.ts:23]: Wait, what does this syntax with the `;` inside the `{}` brackets mean? I've never seen that before 🤔

File: playground/src/hooks/athena_fetcher.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+export function useAthenaFetcher() {
+  const { athenaUrl, athenaSecret, module, moduleConfig } = useBaseInfo();  
+
+  return (
+    async (moduleRoute: string, body?: any) => {
+      if (module === undefined) {
+        return undefined;
+      }
+
+      const url = `${athenaUrl}/modules/${module.type}/${module.name}` + moduleRoute;
```
Comment (pal03377) [inline playground/src/hooks/athena_fetcher.ts]:
```suggestion
      const url = `${athenaUrl}/modules/${module.type}/${module.name}${moduleRoute}`;
```

**Review by pal03377 — COMMENTED** [review]
Nice approach with the hooks! I only got a rough overview of the code (I only read through it once), but I believe it is valuable to us to merge this right now. I only have a few code comments you might want to include:

File: playground/src/hooks/athena/send_feedback.ts (hunk 9 above)
Comment (FelixTJDietrich) [reply playground/src/hooks/athena/send_feedback.ts:23]: You mean 
`{ exercise: Exercise; submission: Submission; feedback: Feedback; }`?
That is just an inline type.

**Comment by FelixTJDietrich** [comment]
Fixed some more type issues, will pay more attention to the checks in the future. Should have checked my mail ^^ 
Thanks for the review @pal03377 :pray: I've implemented your suggestions.

File: playground/src/hooks/athena/send_feedback.ts (hunk 9 above)
Comment (pal03377) [reply playground/src/hooks/athena/send_feedback.ts:23]: Ahh, I see 💡

**Review by pal03377 — APPROVED** [review]
Code looks good. I'll test locally and merge if it works 👍

**Comment by pal03377** [comment]
screenshot-2023-07-24_001680
 

I get this in the console. We already chatted about this shortly in Slack, I'll open this as a new issue

**Comment by pal03377** [comment]
Tested locally using `module_example`, didn't get any other issues apart from the one mentioned above 👍

**Pull request merged**